_posts/2025-03-05-#31405.md

```diff
@@ -7,20 +7,46 @@ authors: [mzumsande]
+	- As soon as a block header becomes part of the fully validated current chain, `ActivateBestChain()` will attempt to connect the block. If this final validation succeeds, the block is connected and its `nStatus` is [raised](https://github.com/bitcoin/bitcoin/blob/3c1f72a36700271c7c1293383549c3be29f28edb/src/validation.cpp#L2762) to the ultimate `BLOCK_VALID_SCRIPTS`.
+
+- `CBlockIndex` holds a `pprev` pointer to its predecessor which makes it is trivial to iterate backwards (towards the genesis block) over the block tree. Iterating forwards (towards the chain tip) is not possible, because a block can have multiple successors pointing to it. The lack of forward iteration makes operations such as finding a new best block header expensive, because we have to iterate over the entire `m_block_index` block tree again.
+
+- `ChainstateManager::m_best_header` and `ActiveChain.Tip()` are both `CBlockIndex` pointers. During normal operation, they'll often converge to the same block with the highest PoW. However, it is crucial to distinguish their meaning and use. `ActiveChain.Tip()` returns the `CBlockIndex*` of the fully validated and connected block that is the tip of the currently active chain. `m_best_header` points to the header with the most PoW that the node knows about. It represents a node's view of what the most-PoW chain might look like, but it is possible that validation won't allow it to progress there. As such, `m_best_header` is just a hint.
 
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. Which purpose(s) does `ChainstateManager::m_best_header` serve?
+
+1. Prior to this PR, which of these statements are true, if any?
+	- A) a `CBlockIndex` with an INVALID predecessor will ALWAYS have a `BLOCK_FAILED_CHILD` `nStatus`
+	- B) a `CBlockIndex` with only VALID predecessors will NEVER have a `BLOCK_FAILED_CHILD` `nStatus`
+
+1. One of the goals of this PR is to ensure `m_best_header`, and the `nStatus` of successors of an invalid block are always correctly set. Which functions are directly responsible for updating these values?
+
+1. For which scenario(s), if any, can commit [validation: call InvalidChainfound also from AcceptBlock](https://github.com/bitcoin-core-review-club/bitcoin/commit/163b01eae82a9dd94dcda94cf13001e46b4a8a56) introduce behaviour change?
+
+1. Most of the logic in commit [validation: in invalidateblock, calculate m_best_header right away](https://github.com/bitcoin-core-review-club/bitcoin/commit/4100495125e9a06b2403f7520fae9f45c3fd9e4c) implements finding the new best header. What prevents us from just using `RecalculateBestHeader()` here? 
```

Comment:
If I understand this correctly...? Just a nit for ease of parsing
```suggestion
1. Most of the logic for finding the new best header is implemented in commit [validation: in invalidateblock, calculate m_best_header right away](https://github.com/bitcoin-core-review-club/bitcoin/commit/4100495125e9a06b2403f7520fae9f45c3fd9e4c). What prevents us from just using `RecalculateBestHeader()` here? 
```

Reply:
I think I prefer my wording - the point is that the commit (re-)implements logic that already exists in a different function. I think this is less clear in your suggestion?

Altneratively, would the below be more clear?

```suggestion
1. Most of the logic in commit [validation: in invalidateblock, calculate m_best_header right away](https://github.com/bitcoin-core-review-club/bitcoin/commit/4100495125e9a06b2403f7520fae9f45c3fd9e4c) is about finding the new best header. What prevents us from just using `RecalculateBestHeader()` here? 
```

Reply:
Just a nit, and perhaps I misunderstood! Feel free to ignore.